### PR TITLE
daemon: defend on minimum recommended kernel version

### PR DESCRIPTION
**Pull request by pkalever**
Do not start gluster-block daemon if the current kernel version doesn't
meet the recommendation.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

### CODE REVIEW

**Comment by pkalever** [comment]
@lxbsz please review.

Thanks!

**Comment by pkalever** [comment]
Note: Upstream kernel version must be chosen, the one used at this patch is just for code building.

@lxbsz please suggest.

Thanks.

File: daemon/gluster-blockd.c [hunk 1]
```diff
@@ -262,13 +265,70 @@ glusterBlockDParseArgs(int count, char **options)
...
+    } else {
+      tptr++;
+    }
+
+    if (i >= VERNUM_BUFLEN) {
+      break;
+    }
+  }
+
+  /* Minimum recommended kernel version: 4.13.13-300.fc27.x86_64 */
```
Comment (lxbsz) [inline daemon/gluster-blockd.c]: Why only the fedora kernel version ? Then the upstream code won't work in the RHEL and CentOS.

**Comment by lxbsz** [comment]
> Note: Upstream kernel version must be chosen, the one used at this patch is just for code building.

Shouldn't we check the Linux distribution type first then the kernel version ?

**Comment by pkalever** [comment]
> > Note: Upstream kernel version must be chosen, the one used at this patch is just for code building.
> 
> Shouldn't we check the Linux distribution type first then the kernel version ?

Hmm. I thought of having different code in upstream and downstream as mentioned at https://bugzilla.redhat.com/show_bug.cgi?id=1594331#c7

But, yeah. It makes-more-sense to have a distro check and forward.

Will add this soon.

Thanks @lxbsz

File: daemon/gluster-blockd.c [hunk 2]
```diff
@@ -262,13 +267,103 @@ glusterBlockDParseArgs(int count, char **options)
...
+    tptr = "3.10.0-862.11.1.el7"; /* Minimum recommended rhel kernel version */
+    if (KERNEL_VERSION(vNum[0], vNum[1], vNum[2]) < KERNEL_VERSION(3, 10, 0)) {
+      goto out;
+    } else if (KERNEL_VERSION(vNum[0], vNum[1], vNum[2]) == KERNEL_VERSION(3, 10, 0)) {
+      if (KERNEL_VERSION(vNum[3], vNum[4], vNum[5]) < KERNEL_VERSION(862, 11, 1)) {
+        goto out;
+      }
+    }
+  } else {
+    LOG("mgmt", GB_LOG_INFO, "%s", "Skipping kernel version check.");
```
[the remark this answers is not in the document]
Comment (pkalever) [reply daemon/gluster-blockd.c]: Sure I think that's doable thing :-)

File: daemon/gluster-blockd.c [hunk 3]
```diff
@@ -262,13 +267,103 @@ glusterBlockDParseArgs(int count, char **options)
...
+      tptr++;
+    }
+
+    if (i >= VERNUM_BUFLEN) {
+      break;
+    }
+  }
+
+  if (strstr(out, "fedora")) {
+    tptr = "4.13.13-300.fc27"; /* Minimum recommended fedora kernel version */
```
[the remark this answers is not in the document]
Comment (pkalever) [reply daemon/gluster-blockd.c]: Sure will adopt this. Thanks for the quick check @lxbsz.

**Comment by pkalever** [comment]
@lxbsz hope all the review comments are resolved.

Thanks!

File: daemon/gluster-blockd.c [hunk 4]
```diff
@@ -262,13 +267,108 @@ glusterBlockDParseArgs(int count, char **options)
...
+      }
+    }
+  } else {
+    LOG("mgmt", GB_LOG_INFO, "Distro %s. Skipping kernel version check.", out);
+  }
+
+  LOG("mgmt", GB_LOG_INFO, "Distro %s. Current kernel version: '%s'.",
+      out, verStr.release);
+
+  return;
```
Comment (lxbsz) [inline daemon/gluster-blockd.c]: Brfore 'return' here, shouldn't we close the fp ?

Comment (pkalever) [reply daemon/gluster-blockd.c]: You got it. Will refresh.

**Comment by lxbsz** [comment]
I looks good to me and test it on rhel node:

 [at dyn-config.c+468 :<glusterBlockDynConfigStart>]
[2018-09-25 09:34:20.738967] INFO: Distro ID="rhel". Current kernel version: '3.10.0-948.el7.test.x86_64.debug'. [at gluster-blockd.c+342 :<gbMinKernelVersionCheck>]
[2018-09-25 09:34:21.688473] INFO: capabilities fetched successfully [at gluster-blockd.c+474 :<main>]

Tested-by: Prasanna Kumar Kalever <xiubli@redhat.com>
Reviewed-by: Prasanna Kumar Kalever <xiubli@redhat.com>

Thanks,

**Comment by pkalever** [comment]
@lxbsz Thanks!

**Pull request merged**